src/Chapter3.hs

```diff
@@ -480,6 +511,45 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
 
+data Castle = MkCastle String deriving (Show)
+data Facility = Church | Library deriving (Show)
+data Inhabitants = One | Two | Three | Four deriving (Show)
+data House = MkHouse Inhabitants deriving (Show)
+
+inhabitantsPerHouse :: House -> Int
+inhabitantsPerHouse (MkHouse One) = 1
+inhabitantsPerHouse (MkHouse Two) = 2
+inhabitantsPerHouse (MkHouse Three) = 3
+inhabitantsPerHouse (MkHouse Four) = 4
+
+inhabitantsPerCity :: [House] -> Int
+inhabitantsPerCity xs = sum $ map inhabitantsPerHouse xs
+
+data City = MkCity
+  { cityCastle :: Maybe Castle
+  , cityFacility :: Facility
+  , cityHouses :: [House]
+  , cityWall :: Bool
+  } deriving (Show)
```

Comment:
This is the correct data type, that can represent what we have in requirements 👍🏼 
However, this also allows having some states that should not ideally exist: like having walls without any castle at all. With this representation, this is possible. However, we actually can restrict this on type-level by providing more representative types 🙂 This is actually the greatest power of Haskell and type system 😍 

This pattern is called "Make illegal states unrepresentable", and you can read more about it here:

* https://kowainik.github.io/posts/haskell-mini-patterns#make-illegal-states-unrepresentable